### ISSUE TITLE
Clarify Tasks wx support

### ISSUE THREAD

**Issue by fbender** [enthought/pyface#219]
Per https://github.com/enthought/pyface/blame/4e454d3da537be74acf6919ecaa87555e6947579/docs/source/overview.rst#L256, Tasks do not support the wx backend. There have been a couple of releases since July 2015. 

Could you please clarify if Tasks support the wx backend by now?

Thanks!

**Comment by jdmarch** [comment]
No, Tasks does not support wx and there are no plans for it to do so.

> Tasks, which is currently available only for Qt, but is under active development.

To be explicit, this does *not* mean that wx support is under active development. It means that tasks is under active development.

**Comment by fbender** [comment]
Thank you for the quick reply. That's very unfortunate. 
There's an existing app I'd like to convert to ETS (and it currently uses wx directly), but this can only happen piece-by-piece so I need to stick with wx for the time being. Workbench it is, then!

@jdmarch are you going to clarify the Docs or should I send in a PR?

**Comment by jdmarch** [comment]
Please feel free to submit a PR. 

Regarding your plans, perhaps others with workbench experience can comment. @corranwebster ?

**Comment by fbender** [comment]
See #220

> Regarding your plans, perhaps others with workbench experience can comment. @corranwebster ?

That would be awesome! I'm having trouble finding good documentation on how to use Workbench via Envisage (I'm new to the ETS stack but absolutely love the concept). I will take a look at the plugins examples in enthought/envisage but I'm not sure how far this will get me, especially considering that the (non-Tasks) examples haven't been updated in a long time … 
FYI: The application is part of the [NANOBED project @ University of Strathclyde](http://craftprospect.com/nanobed/) and features a tree view of the sat, mission & env model plus a few "tabs" containing simulation and analysis "editors" that read and in some cases may modify the tree. So if there's any Open Source application that is similar to this structure and is using the Envisage Workbench API, I'd be glad to learn from it!

**Comment by corranwebster** [comment]
@fbender There was an active effort to port Tasks to Wx by @robmcmullen (see the master branch of his fork of pyface: https://github.com/robmcmullen/pyface).  I don't know the current state of his effort, other than it never made it to a pull-request.  However it might provide a kicking-off point for you, particularly if it does most of what you need it to do in its current state.  If you or your team were willing to 

There's a high-level overview of Workbench in the Envisage documentation here http://docs.enthought.com/envisage/envisage_core_documentation/workbench.html and also some documentation of the "DockWindow" infrastructure of Workbench here https://github.com/enthought/pyface/tree/master/docs 

In terms of examples there is a simple example here https://github.com/enthought/pyface/tree/master/examples/workbench and Envisage-based examples here https://github.com/enthought/envisage/tree/master/examples/plugins/workbench

Just as Tasks can be used without Envisage, the same is true of Workbench, but I have never used it stand-alone, so can't comment directly.

**Comment by fbender** [comment]
That's good to know! 

@robmcmullen if you read this, may you please comment on the status of your fork and if you need any help rebasing and prepare a PR? I could not find a way to contact you directly (other than writing issues in unrelated projects, which I will try to avoid).

I will try to have a look at the links you provided (most were on my list already), thanks. Is there any doc on how to integrate existing `wx.Panel` (and maybe others) into an Envisage/TraitsUI/Pyface app? I only found a way to [include Mayavi into a wx app](http://docs.enthought.com/mayavi/mayavi/auto/example_wx_embedding.html#example-wx-embedding), which I assume works for any TraitsUI panel, but not the other way around.

**Comment by robmcmullen** [comment]
Hi! I do have a wx version of tasks that I'm using in a project [Omnivore](https://github.com/robmcmullen/omnivore) that is under active development. I'm heavily invested in my wx port of tasks because of this.

The last time synced my pyface/traits/traitsui/etc with the master branches, pyface still had a pyface/ui/wx/tasks directory but that disappeared some time ago and I stopped pulling from the pyface source tree.

I'd love to get this integrated back into the tree, but I never prepared a PR because the enthought mailing list died back in 2014 and I didn't think there was continuing interest in wx.

I don't know how far behind I am now, or how much work it would take to bring it up to current versions. I will take a look. I think I was current with the 4.4 tree. The latest tags I have are 4.4 in some libraries, and 4.5rc1 on traits.

If you want to try my versions, you'll have to use wxPython-3.0.2 (Classic, not Phoenix) and python 2.7. I fixed some minor things in all of the libraries, so use all of these instead of the standard Enthough libs:

https://github.com/robmcmullen/traits.git
https://github.com/robmcmullen/pyface.git
https://github.com/robmcmullen/traitsui.git
https://github.com/enthought/apptools.git
https://github.com/robmcmullen/envisage.git

I would love help to figure out what's changed and submit some pull requests with the hopes that it could make it back into the tree, even if an experimental or unsupported version. Long term, I would love to have the resources to port tasks to python 3 if wxPython-Phoenix ever gets a stable, supported version.

**Comment by fbender** [comment]
Awesome! May somebody from Enthought please comment on the pyface/ui/wx/tasks directory?

@robmcmullen I think the first thing to do would be to integrate your lib fixes upstream. I'm new to Python and wx, so I probably can only help in a limited way, but that seems like a doable task, especially if it's just minor fixes. I will try to have a look by end of the coming week. Just let me know what you were able to accomplish in the meantime.

As for the Tasks wx backend, I assume it's a bit more complicated than just dropping in the files and "registering" the backend so I'm not sure if I can help here. You may point me to some small/easy tasks if you want to.

> Long term, I would love to have the resources to port tasks to python 3 if wxPython-Phoenix ever gets a stable, supported version.

Since Enthought is already porting all the code to Python 3, the Tasks wx backend should be compatible with Python 3 when merging – even if wxPython isn't, yet. But that will ease the tansition. As for Phoenix, see wxWidgets/Phoenix#101 – my impression is that Phoenix is coming sooner than later now that wxWidgets 3.1 has been released. There is already a migration guide at http://wxpython.org/Phoenix/docs/html/MigrationGuide.html and http://wxpython.org/Phoenix/docs/html/classic_vs_phoenix.html. I've also seen a some projects already relying on pre-release versions of Phoenix, so it might be an option to port to Phoenix already, or write a small compat layer (for the stuff that isn't in the Phoenix compat layer already) for the thing needed by the backend (if that is feasible at all) so you can just drop in Phoenix once it's stable.

**Comment by jdmarch** [comment]
Any comments, @RobinD42 ? ^

**Comment by corranwebster** [comment]
@robmcmullen @fbender Not sure what, if anything happened to the wx/tasks directory - most likely it got cleaned up by someone, but it's certainly not indicative of a lack of interest in the code.  Tasks has been pretty stable over the last few years, so I imagine it won't be too hard to bring it in: at its simplest, it _should_ be just a matter of dropping the files in the appropriate `pyface/ui/wx/tasks/...` location and the toolkit infrastructure should just pick it up automatically.

There may be some more work reconciling fixes that Rob has made elsewhere, but we're interested in those as well!  I'm travelling this week, so I don't know that I will have the bandwidth to help in the next few days.

I'll let @RobinD42 comment on the wider Wx plans if he has time.

**Comment by robmcmullen** [comment]
@corranwebster I'll see if I can generate some pull requests later this week for the minor stuff in the other libraries, and then see what it takes to get pyface tasks current. Maybe if I drop `pyface/ui/wx/tasks` back in the current version, it will just work.

@fbender as for Phoenix, I'm not going to work on that for a while, as I need an unrelated large library to make it over to python 3. Also, I'll wait until there's a stable Phoenix. The odd dotted releases of wx (e.g. 2.9, 3.1) have historically been development versions, so I'm not going to attempt to start a port to a potentially moving target. :)

**Comment by RobinD42** [comment]
wxPython Phoenix is intended to be mostly compatible with code that runs on wxPython Classic 3.0.3, and where it is different it will usually be due to things like not carrying forward hacks that are no longer needed, etc.  Hopefully most of the differences will be obvious once you've worked through a few of them, and will often be such that changing things to do it the "new" way will actually work on Classic too because it was the "right" way there.  A lot of the differences will have compatibility aliases that raise deprecation warnings, more advanced differences are covered in the migration guide.

I am shooting for having an official alpha of Phoenix in PyPI within a month or two, with support for Python 2.7, 3.4 and 3.5.  It will still be based on wxWidgets 3.0.x in order to be simpler to port to, as moving to 3.1 may break a few additional APIs.  The plan is that once there is an official non-beta release of Phoenix then I will start the effort on moving to wxWidgets 3.1.

**Comment by robmcmullen** [comment]
I've started work merging. `pyface` is the big one (obviously!) so I saved it for last. You've previously merged my minor changes to `traitsui` and `envisage`, and here's the only new code I have for `traits`:

https://github.com/enthought/traits/pull/316

which is a small bit of code to support pulling images from the resource files generated by py2exe & py2app. On to `pyface`!

**Comment by robmcmullen** [comment]
First of two pull requests on pyface is a smaller one where I separated out all the stuff that isn't dependent on tasks.

https://github.com/enthought/pyface/pull/222

Working on the tasks support pull request now.

**Comment by robmcmullen** [comment]
wx support for tasks https://github.com/enthought/pyface/pull/223
